server/comms/middleware.go

```diff
@@ -0,0 +1,121 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package comms
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"decred.org/dcrdex/dex"
+	"decred.org/dcrdex/dex/msgjson"
+	"github.com/go-chi/chi"
+)
+
+type contextKey int
+
+// These are the keys for different types of values stored in a request context.
+const (
+	ctxThing contextKey = iota
+)
+
+// limitRate is rate-limiting middleware that checks whether a request can be
+// fulfilled.
+func (s *Server) limitRate(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		code, err := s.meterIP(r.RemoteAddr)
```

Comment:
When I had played with per IP limits in https://github.com/chappjc/dcrdex/commit/9607307e806275be77ddc313ae956cad439f823a#diff-e7126c606b49bf8ca255171ae2f485e8d92f442af17e953700f49bcea7215507R27, I realized that dealing with IPv6 was a bit trickier because it's really the prefix that you need to consider since the device address is trivially adjusted by the client. https://www.ibm.com/support/knowledgecenter/en/STCMML8/com.ibm.storage.ts3500.doc/opg_3584_IPv4_IPv6_prefix_subnet_mask.

The code I landed on to take an IP address string like RemoteAddr and give a key that masks out the the non-prefix bits of IPv6:

```go
type IP [net.IPv6len]byte

func parseIP(addr string) IP {
	ip := net.ParseIP(addr)
	if ip.To4() == nil {
		ip = ip.Mask(net.CIDRMask(64, 128))
	}
	var ipKey IP
	copy(ipKey[:], ip)
	return ipKey
}
```